test/e2e-monitoring/monitoring_test.go

```diff
@@ -0,0 +1,134 @@
+							For:   &duration,
+							Labels: map[string]string{
+								"bar": "baz",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	gvr := schema.GroupVersionResource{
+		Group:    resource.GroupVersionKind().Group,
+		Version:  resource.GroupVersionKind().Version,
+		Resource: "prometheusrules",
+	}
+
+	// Initialize pre-requisites.
+	cache := resourceapply.NewResourceCache()
+	recorder := events.NewInMemoryRecorder("TestResourceVersionApplication")
+	dynamicClient, err := dynamic.NewForConfig(config)
+	require.NoError(t, err)
+
+	// Create the resource.
+	unstructuredResource := &unstructured.Unstructured{}
+	unstructuredResourceMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&resource)
+	require.NoError(t, err)
+	unstructuredResource.SetUnstructuredContent(unstructuredResourceMap)
+	gotUnstructured, didUpdate, err := resourceapply.ApplyUnstructuredResourceImproved(
+		context.TODO(),
+		dynamicClient,
+		recorder,
+		unstructuredResource,
+		cache,
+		gvr,
+		nil,
+		nil,
+	)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		t.Fatalf("Failed to create resource: %v", err)
+	}
+	require.True(t, didUpdate)
+	existingResource, err := dynamicClient.Resource(gvr).Namespace(resource.GetNamespace()).Get(context.TODO(), resource.GetName(), metav1.GetOptions{})
+	require.NoError(t, err)
+	require.Equal(t, gotUnstructured.UnstructuredContent(), existingResource.UnstructuredContent())
+
+	// Update the resource with a change, without specifying a resource version.
+	resource.Spec.Groups[0].Rules[0].Labels["bar"] = "qux"
+	unstructuredResourceMap, err = runtime.DefaultUnstructuredConverter.ToUnstructured(&resource)
+	require.NoError(t, err)
+	unstructuredResource.SetUnstructuredContent(unstructuredResourceMap)
+	gotUnstructured, didUpdate, err = resourceapply.ApplyUnstructuredResourceImproved(
+		context.TODO(),
+		dynamicClient,
+		recorder,
+		unstructuredResource,
+		cache,
+		gvr,
+		nil,
+		nil,
+	)
+	if err != nil {
+		t.Fatalf("Failed to update resource: %v", err)
+	}
+	require.True(t, didUpdate)
+	existingResource, err = dynamicClient.Resource(gvr).Namespace(resource.GetNamespace()).Get(context.TODO(), resource.GetName(), metav1.GetOptions{})
+	require.NoError(t, err)
+	require.Equal(t, gotUnstructured.UnstructuredContent(), existingResource.UnstructuredContent())
+
+	// Update the resource without any changes, without specifying a resource version.
+	gotUnstructured, didUpdate, err = resourceapply.ApplyUnstructuredResourceImproved(
+		context.TODO(),
+		dynamicClient,
+		recorder,
+		unstructuredResource,
+		cache,
+		gvr,
+		nil,
+		nil,
+	)
+	if err != nil {
+		t.Fatalf("Failed to update resource: %v", err)
+	}
+	require.False(t, didUpdate)
+	existingResource, err = dynamicClient.Resource(gvr).Namespace(resource.GetNamespace()).Get(context.TODO(), resource.GetName(), metav1.GetOptions{})
+	require.NoError(t, err)
+	require.Equal(t, gotUnstructured.UnstructuredContent(), existingResource.UnstructuredContent())
```

Comment:
does `require.Equal` know how to compare `maps` reliably ?

Reply:
instead of comparing to the obj we received from the apply function shouldn't we compare to the expected one ?

Reply:
expected object is the one we created at the beginning of the test (`resource`) which was later updated at line `90`, right ?

Reply:
by comparing against the expected object we know there no other (unexpected) modifications were made.

Reply:
> does require.Equal know how to compare maps reliably ?

ok, it looks like it uses `reflect.DeepEqual` under the hood.

Reply:
> we know there no other (unexpected) modifications were made.

There's a few server-side added fields we'll need to inject, but I wonder if [`12f2bce` (#1833)](https://github.com/openshift/library-go/pull/1833/commits/12f2bce9c9c28af97c00861845613ae2d3f0622d#diff-c44fd9ab87f08101aadd180c482e8f18c866a305b78c7289a95a518cacdf2203) suffices for this requirement?

Reply:
Pushed.

Reply:
I would prefer comparing against the expected object. Now, I think we should compare both the received object and the stored object against the expected one, since callers might use the received object as well, wdyt?

I don't expect many fields to be changed by the server. We could just overwrite or clear the ones we know will be provided by the server, such as RV.

Reply:
Sorry, can you clarify what the `expected`, `received`, and `stored` objects are in this context?

To me (and I'm probably wrong here), `unstructuredResource` seems the same as the `expected` and `stored` object (as it's what we originally create in the beginning, and what we store?), whereas `gotUnstructured` is the one we `receive`?

Reply:
Also, I believe we are comparing the received object with the expected one here: [`17a435d` (#1823)](https://github.com/openshift/library-go/pull/1823/commits/17a435de41df465c17974773c06c77fa9d7fddfe#diff-c44fd9ab87f08101aadd180c482e8f18c866a305b78c7289a95a518cacdf2203R114-R120)? Not sure what the stored object is though (do you mean the one in the cluster?).

Reply:
Pushed a patch that compares the expected (original) resource to the existing (cluster) and the received (returned by `ApplyUnstructuredResourceImproved`).